From f4d8f9e76e17dbf389a0fb4a4109f7623cf26b38 Mon Sep 17 00:00:00 2001
From: "Tao, Jing" <jing.tao@intel.com>
Date: Tue, 5 May 2015 14:13:11 +0800
Subject: [PATCH 378/462] atomisp2: fix panic on acc video node close

The acc video node does not have media
control entity context that will cause
NULL pointer kernel panic by calling
atomisp_subdev_source_pad()

As source_pad is only required on normal
video nodes, the fix is only calling this
function is the video node is not acc
video node.

Change-Id: I956956bda270c53f7feb4dc75726a72ebe9f9689
Tracked-On: IMINAN-35108
Signed-off-by: Tao, Jing <jing.tao@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/54285
Approver: Bin B Han <bin.b.han@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Build: IMIN (SC)
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../drivers/media/pci/atomisp2/atomisp_driver/atomisp_fops.c       | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_fops.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_fops.c
index e97a438..29fea23 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_fops.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_fops.c
@@ -899,7 +899,6 @@ static int atomisp_release(struct file *file)
 	bool acc_node;
 	struct v4l2_requestbuffers req;
 	struct v4l2_subdev_fh fh;
-	uint16_t source_pad = atomisp_subdev_source_pad(vdev);
 	struct v4l2_rect clear_compose = {0};
 	int ret = 0;
 
@@ -1017,11 +1016,13 @@ subdev_uninit:
 		dev_err(isp->dev, "Failed to power off device\n");
 
 done:
-	atomisp_subdev_set_selection(&asd->subdev, &fh,
+	if (!acc_node) {
+		atomisp_subdev_set_selection(&asd->subdev, &fh,
 				V4L2_SUBDEV_FORMAT_ACTIVE,
-				source_pad,
+				atomisp_subdev_source_pad(vdev),
 				V4L2_SEL_TGT_COMPOSE, 0,
 				&clear_compose);
+	}
 	rt_mutex_unlock(&isp->mutex);
 	mutex_unlock(&isp->streamoff_mutex);
 
-- 
1.9.1

